core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -0,0 +1,263 @@
+      if (transport != null) {
+        transport.ping(callback, executor);
+        return;
+      }
+      if (!shutdown) {
+        PendingPing pendingPing = new PendingPing(callback, executor);
+        pendingPings.add(pendingPing);
+        return;
+      }
+    }
+    executor.execute(new Runnable() {
+        @Override public void run() {
+          callback.onFailure(
+              Status.UNAVAILABLE.withDescription("transport shutdown").asException());
+        }
+      });
+  }
+
+  /**
+   * Prevents creating any new streams until {@link #setTransport} is called. Buffered streams are
+   * not failed, so if {@link #shutdown} is called when {@link #setTransport} has not been called,
+   * you still need to call {@link setTransport} to make this transport terminated.
+   */
+  @Override
+  public void shutdown() {
+    boolean shouldCallTerminated = false;
+    synchronized (lock) {
+      if (shutdown) {
+        return;
+      }
+      shutdown = true;
+      if (pendingStreams == null || pendingStreams.isEmpty()) {
+        pendingStreams = null;
+        shouldCallTerminated = true;
+      }
+    }
+    listener.transportShutdown(
+        Status.OK.withDescription("Channel requested transport to shut down"));
+    if (shouldCallTerminated) {
+      listener.transportTerminated();
+    }
+  }
+
+  /**
+   * Shuts down this transport and cancels all streams that it owns, hence immediately terminates
+   * this transport.
+   */
+  public void shutdownNow(Status status) {
+    shutdown();
+    Collection<PendingStream> savedPendingStreams = null;
+    synchronized (lock) {
+      if (pendingStreams != null) {
+        savedPendingStreams = pendingStreams;
+        pendingStreams = null;
+      }
+    }
+    if (savedPendingStreams != null) {
+      for (PendingStream stream : savedPendingStreams) {
+        stream.setError(status);
+      }
+      listener.transportTerminated();
+    }
+    // If savedPendingStreams == null, transportTerminated() has already been called in shutdown().
+  }
+
+  /**
+   * Transfers all the pending and future requests and pings to the given transport.
+   *
+   * <p>May only be called after {@link #start(Listener)}.
+   *
+   * <p>{@code transport} will be used for all future calls to {@link #newStream}, even if this
+   * transport is {@link #shutdown}.
+   */
+  public void setTransport(ClientTransport transport) {
+    boolean savedShutdown;
+    synchronized (lock) {
+      Preconditions.checkState(delegate == null, "setTransport already called");
+      Preconditions.checkState(listener != null, "start() not called");
+      delegate = Preconditions.checkNotNull(transport, "transport");
+      savedShutdown = shutdown;
+      for (PendingPing ping : pendingPings) {
+        ping.createRealPing(transport);
+      }
+      pendingPings = null;
+      if (pendingStreams != null) {
+        for (PendingStream stream : pendingStreams) {
+          stream.createRealStream(transport);
```

Comment:
Done. I actually call transportTerminated() before calling createRealStream(). I think it's fine because at that point we can already consider the pending streams have been transferred to the real transport, so it conforms to the semantics of being terminated.